main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -150,17 +150,37 @@ object Typer extends Phase[ResolvedAst.Root, TypedAst.Root] {
     */
   private def visitDefn(defn: ResolvedAst.Def, assumedTconstrs: List[Ast.TypeConstraint], root: ResolvedAst.Root, classEnv: Map[Symbol.ClassSym, Ast.ClassContext])(implicit flix: Flix): Validation[TypedAst.Def, TypeError] = defn match {
     case ResolvedAst.Def(sym, spec0, exp0) =>
-      typeCheckDecl(spec0, exp0, assumedTconstrs, isMain = sym.isMain, root, classEnv) map {
-        case (spec, exp) => TypedAst.Def(sym, spec, exp)
-      }
+      for {
+        // check the main signature before typechecking the def
+        _ <- checkMain(defn, classEnv)
+        res <- typeCheckDecl(spec0, exp0, assumedTconstrs, root, classEnv)
+        (spec, exp) = res
+      } yield TypedAst.Def(sym, spec, exp)
+  }
+
+  /**
+    * Checks that, if the function is a main function, it has the right type scheme.
+    */
+  private def checkMain(defn: ResolvedAst.Def, classEnv: Map[Symbol.ClassSym, Ast.ClassContext])(implicit flix: Flix): Validation[Unit, TypeError] = {
+    if (!defn.sym.isMain) {
+      return ().toSuccess
+    }
+
+    val mainScheme = Scheme(Nil, Nil, Type.mkImpureArrow(Type.mkArray(Type.Str), Type.Int32))
```

Comment:
I think this should be a nice constant appearing at the top of `Typer.scala` with a small comment.